README.md
# permessage_deflate [![Build status](https://secure.travis-ci.org/faye/permessage-deflate-ruby.svg)](http://travis-ci.org/faye/permessage-deflate-ruby)

Implements the
[permessage-deflate](https://tools.ietf.org/html/draft-ietf-hybi-permessage-compression)
WebSocket protocol extension as a plugin for
[websocket-extensions](https://github.com/faye/websocket-extensions-ruby).

## Installation

```
$ gem install permessage_deflate
```

## Usage

Add the plugin to your extensions:

```rb
require 'websocket/extensions'
require 'permessage_deflate'

exts = WebSocket::Extensions.new
exts.add(PermessageDeflate)
```

The extension can be configured, for example:

```rb
require 'websocket/extensions'
require 'permessage_deflate'

deflate = PermessageDeflate.configure(
  :level => Zlib::BEST_COMPRESSION,
  :max_window_bits => 13
)

exts = WebSocket::Extensions.new
exts.add(deflate)
```

The set of available options can be split into two sets: those that control the
session's compressor for outgoing messages and do not need to be communicated to
the peer, and those that are negotiated as part of the protocol. The settings
only affecting the compressor are described fully in the [Zlib
documentation](http://ruby-doc.org/stdlib-2.1.0/libdoc/zlib/rdoc/Zlib/Deflate.html#method-c-new):

- `:level`: sets the compression level, can be an integer from `0` to `9`, or
  one of the constants `Zlib::NO_COMPRESSION`, `Zlib::BEST_SPEED`,
  `Zlib::BEST_COMPRESSION`, or `Zlib::DEFAULT_COMPRESSION`
- `:mem_level`: sets how much memory the compressor allocates, can be an integer
  from `1` to `9`, or one of the constants `Zlib::MAX_MEM_LEVEL`, or
  `Zlib::DEF_MEM_LEVEL`
- `:strategy`: can be one of the constants `Zlib::FILTERED`,
  `Zlib::HUFFMAN_ONLY`, `Zlib::RLE`, `Zlib::FIXED`, or `Zlib::DEFAULT_STRATEGY`

The other options relate to settings that are negotiated via the protocol and
can be used to set the local session's behaviour and control that of the peer:

- `:no_context_takeover`: if `true`, stops the session reusing a deflate context
  between messages
- `:request_no_context_takeover`: if `true`, makes the session tell the other
  peer not to reuse a deflate context between messages
- `:max_window_bits`: an integer from `8` to `15` inclusive that sets the
  maximum size of the session's sliding window; a lower window size will be used
  if requested by the peer
- `:request_max_window_bits`: an integer from `8` to `15` inclusive to ask the
  other peer to use to set its maximum sliding window size, if supported


CHANGELOG.md
### 0.1.4 / 2017-09-10

- Use `9` instead of `8` as the `windowBits` parameter to zlib, to deal with
  restrictions introduced in zlib v1.2.9

### 0.1.3 / 2016-05-20

- Amend all warnings issued when running with -W2

### 0.1.2 / 2015-11-06

- The server does not send `server_max_window_bits` if the client does not ask
  for it; this works around an issue in Firefox.

### 0.1.1 / 2014-12-18

- Don't allow configure() to be called with unrecognized options

### 0.1.0 / 2014-12-13

- Initial release


spec/spec_helper.rb
require File.expand_path("../../lib/permessage_deflate", __FILE__)

class Message < Struct.new(:data, :rsv1)
end


spec/permessage_deflate/client_session_spec.rb
require "spec_helper"

describe PermessageDeflate::ClientSession do
  let(:ext)       { PermessageDeflate.configure(options) }
  let(:session)   { ext.create_client_session }
  let(:options)   { {} }
  let(:offer)     { session.generate_offer }
  let(:response)  { {} }
  let(:activate)  { session.activate(response) }

  let(:deflate)   { double(:deflate, :deflate => [0x00, 0x00, 0xff, 0xff].pack("C*")) }
  let(:inflate)   { double(:inflate, :inflate => [0x00, 0x00, 0xff, 0xff].pack("C*")) }
  let(:level)     { Zlib::DEFAULT_COMPRESSION }
  let(:mem_level) { Zlib::DEF_MEM_LEVEL }
  let(:strategy)  { Zlib::DEFAULT_STRATEGY }

  let(:message)   { Message.new("hello", true) }

  def process_incoming_message
    session.process_incoming_message(message)
  end

  def process_outgoing_message
    session.process_outgoing_message(message)
  end

  describe "with default options" do
    it "indicates support for client_max_window_bits" do
      expect(offer).to eq("client_max_window_bits" => true)
    end

    describe "with an empty response" do
      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses context takeover and 15 window bits for inflating incoming messages" do
        activate
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end

      it "uses context takeover and 15 window bits for deflating outgoing messages" do
        activate
        expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the response includes server_no_context_takeover" do
      before { response["server_no_context_takeover"] = true }

      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses no context takeover and 15 window bits for inflating incoming messages" do
        activate
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(2).and_return(inflate)
        expect(inflate).to receive(:finish).exactly(2)
        expect(inflate).to receive(:close).exactly(2)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the response includes client_no_context_takeover" do
      before { response["client_no_context_takeover"] = true }

      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses no context takeover and 15 window bits for deflating outgoing messages" do
        activate
        expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, strategy).exactly(2).and_return(deflate)
        expect(deflate).to receive(:finish).exactly(2)
        expect(deflate).to receive(:close).exactly(2)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the response includes server_max_window_bits" do
      before { response["server_max_window_bits"] = 8 }

      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses context takeover and 9 window bits for inflating incoming messages" do
        activate
        expect(Zlib::Inflate).to receive(:new).with(-9).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the response includes invalid server_max_window_bits" do
      before { response["server_max_window_bits"] = 20 }

      it "rejects the response" do
        expect(activate).to be false
      end
    end

    describe "when the response includes client_max_window_bits" do
      before { response["client_max_window_bits"] = 8 }

      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses context takeover and 9 window bits for deflating outgoing messages" do
        activate
        expect(Zlib::Deflate).to receive(:new).with(level, -9, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the response includes invalid client_max_window_bits" do
      before { response["client_max_window_bits"] = 20 }

      it "rejects the response" do
        expect(activate).to be false
      end
    end
  end

  describe "with no_context_takeover" do
    before { options[:no_context_takeover] = true }

    it "sends client_no_context_takeover" do
      expect(offer).to eq(
        "client_no_context_takeover" => true,
        "client_max_window_bits"     => true
      )
    end

    describe "with an empty response" do
      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses no context takeover and 15 window bits for deflating outgoing messages" do
        activate
        expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, strategy).exactly(2).and_return(deflate)
        expect(deflate).to receive(:finish).exactly(2)
        expect(deflate).to receive(:close).exactly(2)
        process_outgoing_message
        process_outgoing_message
      end
    end
  end

  describe "with max_window_bits" do
    before { options[:max_window_bits] = 9 }

    it "sends client_max_window_bits" do
      expect(offer).to eq("client_max_window_bits" => 9)
    end

    describe "with an empty response" do
      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses context takeover and 9 window bits for deflating outgoing messages" do
        activate
        expect(Zlib::Deflate).to receive(:new).with(level, -9, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the response has higher client_max_window_bits" do
      before { response["client_max_window_bits"] = 10 }

      it "rejects the response" do
        expect(activate).to be false
      end
    end

    describe "when the response has lower client_max_window_bits" do
      before { response["client_max_window_bits"] = 8 }

      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses context takeover and 9 window bits for deflating outgoing messages" do
        activate
        expect(Zlib::Deflate).to receive(:new).with(level, -9, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end
  end

  describe "with invalid max_window_bits" do
    before { options[:max_window_bits] = 20 }

    it "raises when generating the offer" do
      expect { offer }.to raise_error(PermessageDeflate::ConfigurationError)
    end
  end

  describe "with request_no_context_takeover" do
    before { options[:request_no_context_takeover] = true }

    it "sends server_no_context_takeover" do
      expect(offer).to eq(
        "client_max_window_bits"     => true,
        "server_no_context_takeover" => true
      )
    end

    describe "with an empty response" do
      it "rejects the response" do
        expect(activate).to be false
      end
    end

    describe "when the response includes server_no_context_takeover" do
      before { response["server_no_context_takeover"] = true }

      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses no context takeover and 15 window bits for inflating incoming messages" do
        activate
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(2).and_return(inflate)
        expect(inflate).to receive(:finish).exactly(2)
        expect(inflate).to receive(:close).exactly(2)
        process_incoming_message
        process_incoming_message
      end
    end
  end

  describe "with request_max_window_bits" do
    before { options[:request_max_window_bits] = 12 }

    it "sends server_max_window_bits" do
      expect(offer).to eq(
        "client_max_window_bits" => true,
        "server_max_window_bits" => 12
      )
    end

    describe "with an empty response" do
      it "rejects the response" do
        expect(activate).to be false
      end
    end

    describe "when the response has higher server_max_window_bits" do
      before { response["server_max_window_bits"] = 13 }

      it "rejects the response" do
        expect(activate).to be false
      end
    end

    describe "when the response has lower server_max_window_bits" do
      before { response["server_max_window_bits"] = 11 }

      it "accepts the response" do
        expect(activate).to be true
      end

      it "uses context takeover and 11 window bits for inflating incoming messages" do
        activate
        expect(Zlib::Inflate).to receive(:new).with(-11).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end
  end

  describe "with invalid request_max_window_bits" do
    before { options[:request_max_window_bits] = 20 }

    it "raises when generating an offer" do
      expect { offer }.to raise_error(PermessageDeflate::ConfigurationError)
    end
  end

  describe "with level" do
    before { options[:level] = Zlib::BEST_SPEED }

    it "sets the level of the deflate stream" do
      activate
      expect(Zlib::Deflate).to receive(:new).with(Zlib::BEST_SPEED, -15, mem_level, strategy).and_return(deflate)
      process_outgoing_message
    end
  end

  describe "with mem_level" do
    before { options[:mem_level] = 5 }

    it "sets the mem_level of the deflate stream" do
      activate
      expect(Zlib::Deflate).to receive(:new).with(level, -15, 5, strategy).and_return(deflate)
      process_outgoing_message
    end
  end

  describe "with strategy" do
    before { options[:strategy] = Zlib::FILTERED }

    it "sets the strategy of the deflate stream" do
      activate
      expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, Zlib::FILTERED).and_return(deflate)
      process_outgoing_message
    end
  end
end


spec/permessage_deflate/server_session_spec.rb
require "spec_helper"

describe PermessageDeflate::ServerSession do
  let(:ext)       { PermessageDeflate.configure(options) }
  let(:offer)     { {} }
  let(:session)   { ext.create_server_session([offer]) }
  let(:options)   { {} }
  let(:response)  { session.generate_response }

  let(:deflate)   { double(:deflate, :deflate => [0x00, 0x00, 0xff, 0xff].pack("C*")) }
  let(:inflate)   { double(:inflate, :inflate => [0x00, 0x00, 0xff, 0xff].pack("C*")) }
  let(:level)     { Zlib::DEFAULT_COMPRESSION }
  let(:mem_level) { Zlib::DEF_MEM_LEVEL }
  let(:strategy)  { Zlib::DEFAULT_STRATEGY }

  let(:message)   { Message.new("hello", true) }

  def process_incoming_message
    session.process_incoming_message(message)
  end

  def process_outgoing_message
    session.process_outgoing_message(message)
  end

  describe "with default options" do
    describe "with an empty offer" do
      it "generates an empty response" do
        expect(response).to eq({})
      end

      it "uses context takeover and 15 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end

      it "uses context takeover and 15 window bits for deflating outgoing messages" do
        response
        expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the offer includes server_no_context_takeover" do
      before { offer["server_no_context_takeover"] = true }

      it "includes server_no_context_takeover in the response" do
        expect(response).to eq("server_no_context_takeover" => true)
      end

      it "uses no context takeover and 15 window bits for deflating outgoing messages" do
        response
        expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, strategy).exactly(2).and_return(deflate)
        expect(deflate).to receive(:finish).exactly(2)
        expect(deflate).to receive(:close).exactly(2)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the offer includes client_no_context_takeover" do
      before { offer["client_no_context_takeover"] = true }

      it "includes client_no_context_takeover in the response" do
        expect(response).to eq("client_no_context_takeover" => true)
      end

      it "uses no context takeover and 15 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(2).and_return(inflate)
        expect(inflate).to receive(:finish).exactly(2)
        expect(inflate).to receive(:close).exactly(2)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the offer includes server_max_window_bits" do
      before { offer["server_max_window_bits"] = 13 }

      it "includes server_max_window_bits in the response" do
        expect(response).to eq("server_max_window_bits" => 13)
      end

      it "uses context takeover and 13 window bits for deflating outgoing messages" do
        response
        expect(Zlib::Deflate).to receive(:new).with(level, -13, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the offer includes invalid server_max_window_bits" do
      before { offer["server_max_window_bits"] = 20 }

      it "does not create a session" do
        expect(session).to be_nil
      end
    end

    describe "when the offer includes client_max_window_bits" do
      before { offer["client_max_window_bits"] = true }

      it "does not include a client_max_window_bits hint in the response" do
        expect(response).to eq({})
      end

      it "uses context takeover and 15 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the offer includes a client_max_window_bits hint" do
      before { offer["client_max_window_bits"] = 13 }

      it "includes a client_max_window_bits hint in the response" do
        expect(response).to eq("client_max_window_bits" => 13)
      end

      it "uses context takeover and 13 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-13).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the offer includes invalid client_max_window_bits" do
      before { offer["client_max_window_bits"] = 20 }

      it "does not create a session" do
        expect(session).to be_nil
      end
    end
  end

  describe "with no_context_takeover" do
    before { options[:no_context_takeover] = true }

    describe "with an empty offer" do
      it "includes server_no_context_takeover in the response" do
        expect(response).to eq("server_no_context_takeover" => true)
      end

      it "uses no context takeover and 15 window bits for deflating outgoing messages" do
        response
        expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, strategy).exactly(2).and_return(deflate)
        expect(deflate).to receive(:finish).exactly(2)
        expect(deflate).to receive(:close).exactly(2)
        process_outgoing_message
        process_outgoing_message
      end
    end
  end

  describe "with max_window_bits" do
    before { options[:max_window_bits] = 12 }

    describe "with an empty offer" do
      it "does not include server_max_window_bits in the response" do
        expect(response).to eq({})
      end

      it "uses context takeover and 12 window bits for deflating outgoing messages" do
        response
        expect(Zlib::Deflate).to receive(:new).with(level, -12, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the offer has higher server_max_window_bits" do
      before { offer["server_max_window_bits"] = 13 }

      it "includes server_max_window_bits in the response" do
        expect(response).to eq("server_max_window_bits" => 12)
      end

      it "uses context takeover and 12 window bits for deflating outgoing messages" do
        response
        expect(Zlib::Deflate).to receive(:new).with(level, -12, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end

    describe "when the offer has lower server_max_window_bits" do
      before { offer["server_max_window_bits"] = 11 }

      it "includes server_max_window_bits in the response" do
        expect(response).to eq("server_max_window_bits" => 11)
      end

      it "uses context takeover and 11 window bits for deflating outgoing messages" do
        response
        expect(Zlib::Deflate).to receive(:new).with(level, -11, mem_level, strategy).exactly(1).and_return(deflate)
        process_outgoing_message
        process_outgoing_message
      end
    end
  end

  describe "with request_no_context_takeover" do
    before { options[:request_no_context_takeover] = true }

    describe "with an empty offer" do
      it "includes client_no_context_takeover in the response" do
        expect(response).to eq("client_no_context_takeover" => true)
      end

      it "uses no context takeover and 15 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(2).and_return(inflate)
        expect(inflate).to receive(:finish).exactly(2)
        expect(inflate).to receive(:close).exactly(2)
        process_incoming_message
        process_incoming_message
      end
    end
  end

  describe "with request_max_window_bits" do
    before { options[:request_max_window_bits] = 11 }

    describe "with an empty offer" do
      it "does not include client_max_window_bits in the response" do
        expect(response).to eq({})
      end

      it "uses context takeover and 15 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-15).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the offer includes client_max_window_bits" do
      before { offer["client_max_window_bits"] = true }

      it "includes client_max_window_bits in the response" do
        expect(response).to eq("client_max_window_bits" => 11)
      end

      it "uses context takeover and 11 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-11).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the offer has higher client_max_window_bits" do
      before { offer["client_max_window_bits"] = 12 }

      it "includes client_max_window_bits in the response" do
        expect(response).to eq("client_max_window_bits" => 11)
      end

      it "uses context takeover and 11 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-11).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end

    describe "when the offer has lower client_max_window_bits" do
      before { offer["client_max_window_bits"] = 10 }

      it "includes client_max_window_bits in the response" do
        expect(response).to eq("client_max_window_bits" => 10)
      end

      it "uses context takeover and 10 window bits for inflating incoming messages" do
        response
        expect(Zlib::Inflate).to receive(:new).with(-10).exactly(1).and_return(inflate)
        process_incoming_message
        process_incoming_message
      end
    end
  end

  describe "with level" do
    before { options[:level] = Zlib::BEST_SPEED }

    it "sets the level of the deflate stream" do
      response
      expect(Zlib::Deflate).to receive(:new).with(Zlib::BEST_SPEED, -15, mem_level, strategy).and_return(deflate)
      process_outgoing_message
    end
  end

  describe "with mem_level" do
    before { options[:mem_level] = 5 }

    it "sets the mem_level of the deflate stream" do
      response
      expect(Zlib::Deflate).to receive(:new).with(level, -15, 5, strategy).and_return(deflate)
      process_outgoing_message
    end
  end

  describe "with strategy" do
    before { options[:strategy] = Zlib::FILTERED }

    it "sets the strategy of the deflate stream" do
      response
      expect(Zlib::Deflate).to receive(:new).with(level, -15, mem_level, Zlib::FILTERED).and_return(deflate)
      process_outgoing_message
    end
  end
end


lib/permessage_deflate.rb
require 'zlib'

class PermessageDeflate
  root = File.expand_path('..', __FILE__)
  require root + '/permessage_deflate/session'
  require root + '/permessage_deflate/client_session'
  require root + '/permessage_deflate/server_session'

  ConfigurationError = Class.new(ArgumentError)

  VALID_OPTIONS = [
    :level,
    :mem_level,
    :strategy,
    :no_context_takeover,
    :max_window_bits,
    :request_no_context_takeover,
    :request_max_window_bits
  ]

  def self.validate_options(options, valid_keys)
    options.keys.each do |key|
      unless valid_keys.include?(key)
        raise ConfigurationError, "Unrecognized option: #{key.inspect}"
      end
    end
  end

  module Extension
    define_method(:name) { 'permessage-deflate' }
    define_method(:type) { 'permessage' }
    define_method(:rsv1) { true  }
    define_method(:rsv2) { false }
    define_method(:rsv3) { false }

    def configure(options)
      @options ||= nil

      PermessageDeflate.validate_options(options, VALID_OPTIONS)
      options = (@options || {}).merge(options)
      PermessageDeflate.new(options)
    end

    def create_client_session
      ClientSession.new(@options || {})
    end

    def create_server_session(offers)
      offers.each do |offer|
        return ServerSession.new(@options || {}, offer) if ServerSession.valid_params?(offer)
      end
      nil
    end
  end

  include Extension
  extend  Extension

  def initialize(options)
    @options = options
  end
end


lib/permessage_deflate/client_session.rb
class PermessageDeflate
  class ClientSession < Session

    def self.valid_params?(params)
      return false unless super

      if params.has_key?('client_max_window_bits')
        return false unless VALID_WINDOW_BITS.include?(params['client_max_window_bits'])
      end

      true
    end

    def generate_offer
      offer = {}

      if @accept_no_context_takeover
        offer['client_no_context_takeover'] = true
      end

      if @accept_max_window_bits
        unless VALID_WINDOW_BITS.include?(@accept_max_window_bits)
          raise ConfigurationError, 'Invalid value for max_window_bits'
        end
        offer['client_max_window_bits'] = @accept_max_window_bits
      else
        offer['client_max_window_bits'] = true
      end

      if @request_no_context_takeover
        offer['server_no_context_takeover'] = true
      end

      if @request_max_window_bits
        unless VALID_WINDOW_BITS.include?(@request_max_window_bits)
          raise ConfigurationError, 'Invalid value for request_max_window_bits'
        end
        offer['server_max_window_bits'] = @request_max_window_bits
      end

      offer
    end

    def activate(params)
      return false unless ClientSession.valid_params?(params)

      if @accept_max_window_bits and params['client_max_window_bits']
        return false if params['client_max_window_bits'] > @accept_max_window_bits
      end

      if @request_no_context_takeover and !params['server_no_context_takeover']
        return false
      end

      if @request_max_window_bits
        return false unless params['server_max_window_bits']
        return false if params['server_max_window_bits'] > @request_max_window_bits
      end

      @own_context_takeover = !(@accept_no_context_takeover || params['client_no_context_takeover'])
      @own_window_bits = [
        @accept_max_window_bits || MAX_WINDOW_BITS,
        params['client_max_window_bits'] || MAX_WINDOW_BITS
      ].min

      @peer_context_takeover = !params['server_no_context_takeover']
      @peer_window_bits = params['server_max_window_bits'] || MAX_WINDOW_BITS

      true
    end

  end
end


lib/permessage_deflate/server_session.rb
class PermessageDeflate
  class ServerSession < Session

    def self.valid_params?(params)
      return false unless super

      if params.has_key?('client_max_window_bits')
        return false unless ([true] + VALID_WINDOW_BITS).include?(params['client_max_window_bits'])
      end

      true
    end

    def initialize(options, params)
      super(options)
      @params = params
    end

    def generate_response
      response = {}

      # https://tools.ietf.org/html/rfc7692#section-7.1.1.1

      @own_context_takeover = !@accept_no_context_takeover &&
                              !@params['server_no_context_takeover']

      response['server_no_context_takeover'] = true unless @own_context_takeover

      # https://tools.ietf.org/html/rfc7692#section-7.1.1.2

      @peer_context_takeover = !@request_no_context_takeover &&
                               !@params['client_no_context_takeover']

      response['client_no_context_takeover'] = true unless @peer_context_takeover

      # https://tools.ietf.org/html/rfc7692#section-7.1.2.1

      @own_window_bits = [ @accept_max_window_bits || MAX_WINDOW_BITS,
                           @params['server_max_window_bits'] || MAX_WINDOW_BITS
                         ].min

      # In violation of the spec, Firefox closes the connection if it does not
      # send server_max_window_bits but the server includes this in its response
      if @own_window_bits < MAX_WINDOW_BITS and @params['server_max_window_bits']
        response['server_max_window_bits'] = @own_window_bits
      end

      # https://tools.ietf.org/html/rfc7692#section-7.1.2.2

      if client_max = @params['client_max_window_bits']
        client_max = MAX_WINDOW_BITS if client_max == true
        @peer_window_bits = [@request_max_window_bits || MAX_WINDOW_BITS, client_max].min
      else
        @peer_window_bits = MAX_WINDOW_BITS
      end

      if @peer_window_bits < MAX_WINDOW_BITS
        response['client_max_window_bits'] = @peer_window_bits
      end

      response
    end

  end
end


lib/permessage_deflate/session.rb
class PermessageDeflate
  class Session

    VALID_PARAMS = [
      'server_no_context_takeover',
      'client_no_context_takeover',
      'server_max_window_bits',
      'client_max_window_bits'
    ]

    MIN_WINDOW_BITS   = 9
    MAX_WINDOW_BITS   = 15
    VALID_WINDOW_BITS = [8, 9, 10, 11, 12, 13, 14, 15]

    def self.valid_params?(params)
      return false unless params.keys.all? { |k| VALID_PARAMS.include?(k) }
      return false if params.values.grep(Array).any?

      if params.has_key?('server_no_context_takeover')
        return false unless params['server_no_context_takeover'] == true
      end

      if params.has_key?('client_no_context_takeover')
        return false unless params['client_no_context_takeover'] == true
      end

      if params.has_key?('server_max_window_bits')
        return false unless VALID_WINDOW_BITS.include?(params['server_max_window_bits'])
      end

      true
    end

    def initialize(options)
      @level     = options.fetch(:level,     Zlib::DEFAULT_COMPRESSION)
      @mem_level = options.fetch(:mem_level, Zlib::DEF_MEM_LEVEL)
      @strategy  = options.fetch(:strategy,  Zlib::DEFAULT_STRATEGY)

      @accept_no_context_takeover  = options.fetch(:no_context_takeover, false)
      @accept_max_window_bits      = options.fetch(:max_window_bits, nil)
      @request_no_context_takeover = options.fetch(:request_no_context_takeover, false)
      @request_max_window_bits     = options.fetch(:request_max_window_bits, nil)

      @deflate = @inflate = nil
    end

    def process_incoming_message(message)
      return message unless message.rsv1

      inflate = get_inflate

      message.data = inflate.inflate(message.data) +
                     inflate.inflate([0x00, 0x00, 0xff, 0xff].pack('C*'))

      free(inflate) unless @inflate
      message
    end

    def process_outgoing_message(message)
      deflate = get_deflate

      message.data = deflate.deflate(message.data, Zlib::SYNC_FLUSH)[0...-4]
      message.rsv1 = true

      free(deflate) unless @deflate
      message
    end

    def close
      free(@inflate)
      @inflate = nil

      free(@deflate)
      @deflate = nil
    end

  private

    def free(codec)
      return if codec.nil?
      codec.finish rescue nil
      codec.close
    end

    def get_inflate
      return @inflate if @inflate

      window_bits = [@peer_window_bits, MIN_WINDOW_BITS].max
      inflate     = Zlib::Inflate.new(-window_bits)

      @inflate = inflate if @peer_context_takeover
      inflate
    end

    def get_deflate
      return @deflate if @deflate

      window_bits = [@own_window_bits, MIN_WINDOW_BITS].max
      deflate     = Zlib::Deflate.new(@level, -window_bits, @mem_level, @strategy)

      @deflate = deflate if @own_context_takeover
      deflate
    end

  end
end
